File: src/main/com/intellij/lang/jsgraphql/ide/introspection/GraphQLRerunLatestIntrospectionAction.java
/*
 * Copyright (c) 2018-present, Jim Kynde Meyer
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
package com.intellij.lang.jsgraphql.ide.introspection;

import com.intellij.icons.AllIcons;
import com.intellij.openapi.actionSystem.AnAction;
import com.intellij.openapi.actionSystem.AnActionEvent;

public class GraphQLRerunLatestIntrospectionAction extends AnAction {

    private static final String TEXT = "Update the Local Schema by Running Introspection Query";

    public GraphQLRerunLatestIntrospectionAction() {
        super(TEXT, "Re-runs the latest introspection query that was performed to update the local schema, e.g. when a remote schema has been changed", AllIcons.Actions.Rerun);
    }

    @Override
    public void update(AnActionEvent e) {
        super.update(e);
        boolean enabled = false;
        e.getPresentation().setText(TEXT);
        if (e.getProject() != null) {
            final GraphQLIntrospectionTask latestIntrospection = GraphQLIntrospectionService.getInstance(e.getProject()).getLatestIntrospection();
            if (latestIntrospection != null) {
                enabled = true;
                e.getPresentation().setText(TEXT + " (" + latestIntrospection.getEndpoint().getName() + ")");
            }
        }
        e.getPresentation().setEnabled(enabled);
    }

    @Override
    public void actionPerformed(AnActionEvent e) {
        if (e.getProject() != null) {
            GraphQLIntrospectionTask latestIntrospection = GraphQLIntrospectionService.getInstance(e.getProject()).getLatestIntrospection();
            if (latestIntrospection != null) {
                latestIntrospection.getRunnable().run();
            }
        }
    }
}

File: src/main/com/intellij/lang/jsgraphql/ide/introspection/GraphQLRerunLatestIntrospectionAction.kt
/*
 * Copyright (c) 2018-present, Jim Kynde Meyer
 * All rights reserved.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
package com.intellij.lang.jsgraphql.ide.introspection

import com.intellij.lang.jsgraphql.GraphQLBundle
import com.intellij.openapi.actionSystem.ActionUpdateThread
import com.intellij.openapi.actionSystem.AnAction
import com.intellij.openapi.actionSystem.AnActionEvent

class GraphQLRerunLatestIntrospectionAction : AnAction() {
  companion object {
    const val ACTION_ID = "GraphQLRerunLatestIntrospection"
  }

  override fun update(e: AnActionEvent) {
    val project = e.project ?: return
    var enabled = false
    val latestIntrospection = GraphQLIntrospectionService.getInstance(project).latestIntrospection
    if (latestIntrospection != null) {
      enabled = true
      e.presentation.text =
        GraphQLBundle.message("graphql.action.rerun.latest.introspection.schema.title", latestIntrospection.endpoint.displayName)
    }
    e.presentation.isEnabled = enabled
  }

  override fun getActionUpdateThread(): ActionUpdateThread {
    return ActionUpdateThread.BGT
  }

  override fun actionPerformed(e: AnActionEvent) {
    val project = e.project ?: return
    GraphQLIntrospectionService.getInstance(project).latestIntrospection?.runnable?.run()
  }
}
